[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74608
* #74607
* #74606
* __->__ #74226

Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

mikaylagawarecki committed:
Use ScatterGatherKernel for scatter_reduce

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce, *, bool include_input=True)`

- Update `scatter_reduce` to call into the cpu/cuda kernels for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Add an argument `include_input` which indicates whether the value in the `self` Tensor at a given position is included in the reduction with the elements from `src` scattered to that position. For
`I_self = {all indices of self}`
`I_src= {all indices of src}`
`S = {indices of self modified by scatter}`
`self_indices_to_src_indices : I_self --> I_src` maps indices in `self` to a tuple of indices in `src` scattered to that index of `self`
Then  for `s ∈ S` and `t ∈ I\S` when `include_input=False`
`self[s] = reduction_op(src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`
and when `include_input=True` (regular scatter(reduce=op) behavior)
`self[s] = reduction_op(self[s], src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`

The [`optional_out` case of pytorch_scatter.scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L32) can then be handled by 
`torch.zeros(shape).scatter_reduce_(dim, index, src, reduce, include_input=False)`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce, *, bool include_input=True)`

- Update `scatter_reduce` to call into the cpu/cuda kernels for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Add an argument `include_input` which indicates whether the value in the `self` Tensor at a given position is included in the reduction with the elements from `src` scattered to that position. For
`I_self = {all indices of self}`
`I_src= {all indices of src}`
`S = {indices of self modified by scatter}`
`self_indices_to_src_indices : I_self --> I_src` maps indices in `self` to a tuple of indices in `src` scattered to that index of `self`
Then  for `s ∈ S` and `t ∈ I\S` when `include_input=False`
`self[s] = reduction_op(src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`
and when `include_input=True` (regular scatter(reduce=op) behavior)
`self[s] = reduction_op(self[s], src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`

The [`optional_out` case of pytorch_scatter.scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L32) can then be handled by 
`torch.zeros(shape).scatter_reduce_(dim, index, src, reduce, include_input=False)`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce, *, bool include_input=True)`

- Update `scatter_reduce` to call into the cpu/cuda kernels for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Add an argument `include_input` which indicates whether the value in the `self` Tensor at a given position is included in the reduction with the elements from `src` scattered to that position. For
`I_self = {all indices of self}`
`I_src= {all indices of src}`
`S = {indices of self modified by scatter}`
`self_indices_to_src_indices : I_self --> I_src` maps indices in `self` to a tuple of indices in `src` scattered to that index of `self`
Then  for `s ∈ S` and `t ∈ I\S` when `include_input=False`
`self[s] = reduction_op(src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`
and when `include_input=True` (regular scatter(reduce=op) behavior)
`self[s] = reduction_op(self[s], src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`

The [`optional_out` case of pytorch_scatter.scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L32) can then be handled by 
`torch.zeros(shape).scatter_reduce_(dim, index, src, reduce, include_input=False)`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce, *, bool include_input=True)`

- Update `scatter_reduce` to call into the cpu/cuda kernels for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Add an argument `include_input` which indicates whether the value in the `self` Tensor at a given position is included in the reduction with the elements from `src` scattered to that position. For
`I_self = {all indices of self}`
`I_src= {all indices of src}`
`S = {indices of self modified by scatter}`
`self_indices_to_src_indices : I_self --> I_src` maps indices in `self` to a tuple of indices in `src` scattered to that index of `self`
Then  for `s ∈ S` and `t ∈ I\S` when `include_input=False`
`self[s] = reduction_op(src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`
and when `include_input=True` (regular scatter(reduce=op) behavior)
`self[s] = reduction_op(self[s], src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`

The [`optional_out` case of pytorch_scatter.scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L32) can then be handled by 
`torch.zeros(shape).scatter_reduce_(dim, index, src, reduce, include_input=False)`

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/cuda/ScatterGatherKernel.cu:
```diff
@@ -244,6 +262,66 @@ struct cuda_scatter_gather_base_kernel {
+    auto index_stride = self_dim_stride;
+
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(
```
If the only difference between this kernel and cuda_scatter_gather_base_kernel is 
`AT_DISPATCH_FLOATING_TYPES_AND2` vs `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3` it might be easier to explicitly check for a specific scalar type and error out in a unified kernel than copy-pasting the code and just changing this line.

The error message to throw can be found in the [dispatch macro definitions](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L729).

cpuhrsch commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -106,10 +106,9 @@
     ("aten::grid_sampler_3d_backward", datetime.date(9999, 1, 1)),
     ("aten::_transform_bias_rescale_qkv", datetime.date(9999, 1, 1)),
-    ("aten::_scatter_reduce.two", datetime.date(9999, 1, 1)),
+    ("aten::scatter_reduce.two", datetime.date(9999, 1, 1)),
```
Why this change?

mikaylagawarecki replied:
Previously this was here when `_scatter_reduce` was moved to  `scatter_reduce`. Now I changed the signature of `scatter_reduce` so I assumed adding it to this dict was necessary

cpuhrsch commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -15680,16 +15686,9 @@ def ref_pairwise_distance(input1, input2):
     ),
     OpInfo(
         'scatter_reduce',
-        dtypes=all_types_and(torch.float16, torch.bfloat16),
```
How come this was working beforehand?

mikaylagawarecki replied:
Previously this was enabled for cpu only, I sent a longer discussion regarding this offline

cpuhrsch reviewed (changes requested):
Looks pretty good, but I'd try to avoid the code duplication of copy-pasting `cpu_scatter_gather_base_kernel` to deal with the different set of dtypes.

mikaylagawarecki commented on test/test_torch.py:
```diff
@@ -3297,6 +3297,56 @@ def test_scatter_add_bool(self, device):
                                             [False, True, False, True, False],
                                             [True, False, True, False, True]], device=device))
 
+    # FIXME: move to test_scatter_gather_ops.py
```
Sorry about that, previously this test was in the class `TestTorch` which does not allow parameterization with @dtypes because it is not instantiated with `instantiate_device_type_tests(TestTorch, globals())`. Moved this here into the class `TestTorchDeviceType` which allows parameterization by @dtypes. Is there another way around this?

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce, *, bool include_input=True)`

- Update `scatter_reduce` to call into the cpu/cuda kernels for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Add an argument `include_input` which indicates whether the value in the `self` Tensor at a given position is included in the reduction with the elements from `src` scattered to that position. For
`I_self = {all indices of self}`
`I_src= {all indices of src}`
`S = {indices of self modified by scatter}`
`self_indices_to_src_indices : I_self --> I_src` maps indices in `self` to a tuple of indices in `src` scattered to that index of `self`
Then  for `s ∈ S` and `t ∈ I\S` when `include_input=False`
`self[s] = reduction_op(src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`
and when `include_input=True` (regular scatter(reduce=op) behavior)
`self[s] = reduction_op(self[s], src[self_indices_to_src_indices[s]])`
`self[t] = self[t]`

The [`optional_out` case of pytorch_scatter.scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L32) can then be handled by 
`torch.zeros(shape).scatter_reduce_(dim, index, src, reduce, include_input=False)`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -333,6 +527,8 @@ void scatter_scalar_reduce_cpu_kernel(const Tensor& self, const int64_t dim, con
     cpu_scatter_gather_base_kernel<>()(self, dim, index, value,
                                        "scatter_scalar_reduce_multiply_", reduce_multiply);
     break;
+  default:
```
Why did you add this?

mikaylagawarecki replied:
Without this the build on CI will fail. I think it's due to this 

> Using the -Werror compiler flag, a switch statement over a value of an enum type without a default label will fail to compile if any enumerator of the enum doesn’t have a corresponding case. This is sometimes called an exhaustive or defaultless switch statement.

cpuhrsch commented on test/test_scatter_gather_ops.py:
```diff
@@ -158,6 +179,46 @@ def test_scatter_add_mult_index_base(self, device, dtype):
         self.assertEqual(res1[:, 0], n * torch.ones(m, device=device, dtype=dtype), atol=0, rtol=0)
 
+    @onlyCPU
+    @toleranceOverride({torch.float16: tol(atol=1e-2, rtol=0),
```
How did you determine these tolerances?

mikaylagawarecki replied:
For float16 this is copied from [the toleranceOverride decorator on test_scatter__reductions in this file](https://github.com/pytorch/pytorch/blob/22a3a17ea0d61660b53f6642fe22529a7e4e2649/test/test_scatter_gather_ops.py#L135). I copied this for bfloat16 but I'm not sure what tolerance should be used. How does one normally determine this

cpuhrsch replied:
There are default tolerances set, I'd first try running this without an override

mikaylagawarecki replied:
Removed these decorators, thanks!

cpuhrsch reviewed (approved):
Looks great, just added two small comments

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35222842).

mikaylagawarecki committed:
Update on "[BC-breaking] Use ScatterGatherKernel for scatter_reduce (CPU-only)"


Update signature of `scatter_reduce_` to match `scatter_/scatter_add_`

`Tensor.scatter_reduce_(int64 dim, Tensor index, Tensor src, str reduce)`

- Add new reduction options in ScatterGatherKernel.cpp and update `scatter_reduce` to call into the cpu kernel for `scatter.reduce`
- `scatter_reduce` now has the same shape constraints as `scatter_` and `scatter_add_`
- Migrate `test/test_torch.py:test_scatter_reduce` to `test/test_scatter_gather_ops.py`

Differential Revision: [D35222842](https://our.internmc.facebook.com/intern/diff/D35222842)

[ghstack-poisoned]
